### PR TITLE
[SNOW-1482588] Eliminate cast expression for numeric value

### PR DESCRIPTION
**Pull request by sfc-gh-yzou**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1482588

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [x] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [x] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

Today, we generate an extra cast expression for all literal expression to make sure the expression generated is casted to an expected type. However, such cast introduces following problems:
1) In general expression, the final dtype should be decided by snowflake, and user should cast the value explicitly if needed
2) extra cast expression increases expression depth which causes extra overhead in both compilation and execution.  In our experiment with IN expression, the difference can be dramatic.

In this pr, the following changes are applied:
1) For binary expression, multiExpression (used in column in), InExpression, avoid generating sql value with casting for numerical types. For example: 'A' > 1 :: INT will become 'A' > 1, and 'A' in (1 :: INT, 2 :: INT) will be come 'A' in (1, 2). We only do that for expressions, so for column select with value expression, they still follows our python to snowflake mapping.
2) rename to_sql_avoid_offset to to_sql_try_avoid cast to be more clear. 
3) added parameter _eliminate_numeric_sql_value_cast_enabled to control the change.

### CODE REVIEW

File: tests/integ/test_column_names.py
Comment (sfc-gh-yzou) [inline tests/integ/test_column_names.py:846]: @sfc-gh-aling @sfc-gh-stan This case seems returning False for local testing mode, is that expected?

File: tests/integ/test_column_names.py
Comment (sfc-gh-yzou) [inline tests/integ/test_column_names.py:807]: @sfc-gh-aling @sfc-gh-stan The bitwise operator seems returning Boolean type for local testing mode here, is that expected?

File: tests/integ/test_column_names.py
Comment (sfc-gh-yzou) [inline tests/integ/test_column_names.py:347]: @sfc-gh-jdu for this case,  it actually seems should be ArrayType(LongType), but without local testing mode, we seems getting ArrayType(StringType()), do you know if that is what we expected today?

File: tests/integ/test_column_names.py
Comment (sfc-gh-stan) [reply tests/integ/test_column_names.py:846]: Can you elaborate please? What's not being handled correctly?

File: tests/integ/test_column_names.py
Comment (sfc-gh-jdu) [reply tests/integ/test_column_names.py:347]: Yes, because inner type of an array is not enforced, because it's not a structured type. So the server will just return StringType

File: CHANGELOG.md
Comment (sfc-gh-jdu) [inline CHANGELOG.md:11]: We can mention performance improvement for binary column operations and dataframe._in method (users don't know about internal expressions). Also, we should mention the public parameter, like "this performance optimization can be enabled by setting `session.eliminate_numeric_sql_value_cast_enabled=True`"

File: tests/integ/test_column_names.py
Comment (sfc-gh-stan) [reply tests/integ/test_column_names.py:807]: No it's not. I know why though. This is likely the `ColumnEmulator` doesn't have a numeric `dtype` when the bitwise operation happened. If the dtype is `object`, the return value will be `boolean`. That's why we have https://github.com/snowflakedb/snowpark-python/blob/8a4276387e09947d55139fac6a1d6e3590f3f891/src/snowflake/snowpark/mock/_util.py#L215

File: src/snowflake/snowpark/_internal/analyzer/analyzer.py
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/analyzer.py:731]: 👍

File: tests/integ/scala/test_literal_suite.py
Comment (sfc-gh-jdu) [inline tests/integ/scala/test_literal_suite.py:33]: remove?

File: src/snowflake/snowpark/session.py
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/session.py:582]: we should have a set function?

File: tests/integ/test_column_names.py
Comment (sfc-gh-yzou) [reply tests/integ/test_column_names.py:846]: Hi @sfc-gh-stan sorry for the late reply. For this test case, it has a dataframe with one single column named 'INF', and have value 'inf', which is  counted as np.inf, and the test is simply checking if df1["'INF'"] == math.inf, which should be True, Snowpark python seems returning the correct result, but with local testing the result seems False.

File: tests/integ/test_column_names.py
Comment (sfc-gh-yzou) [reply tests/integ/test_column_names.py:807]: Oh, do we have a plan to fix that? is here a ticket to track this ?

Comment (sfc-gh-stan) [reply tests/integ/test_column_names.py:807]: I don't think this is a known issue. @sfc-gh-aling can you help unblock Yun here?

File: tests/integ/test_column_names.py
Comment (sfc-gh-stan) [reply tests/integ/test_column_names.py:846]: No worries, what is `df1["'INF'"]` returning for local testing then? It should also be `np.inf`.

File: tests/integ/test_column_names.py
Comment (sfc-gh-yzou) [reply tests/integ/test_column_names.py:807]: I don't think i am necessary blocking here, I can create a ticket for you to look into this if this is not a known issue, and left a comment on the test

File: tests/integ/scala/test_literal_suite.py
Comment (sfc-gh-yzou) [reply tests/integ/scala/test_literal_suite.py]: rmoved

File: src/snowflake/snowpark/session.py
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/session.py:582]: yes, added, aslo added a telemetry on variable setting

File: tests/integ/test_column_names.py
Comment (sfc-gh-yzou) [reply tests/integ/test_column_names.py:846]: Ok, i think i know what is happening there. when we create a dataframe like this 
```
session.create_dataframe([["inf"]], schema=["'INF'"])
```
the column is actually mapped to string column, in snowflake, when comparing with 'inf' = 'INF'::FLOAT, it was able to do property casting, and return True. However, in local testing i believe it is not able to do such casting, in python, string to numeric value is not castable. That is the behavior difference between python and snowflake

File: tests/integ/test_column_names.py
Comment (sfc-gh-yzou) [reply tests/integ/test_column_names.py:347]: I see, so that is an expected behavior with snowpark on snowflake then.

File: tests/integ/test_session.py
Comment (sfc-gh-stan) [inline tests/integ/test_session.py:687]: This should be an xfail instead. We distinguish between xfail and skipif such that xfail is for tests that will not be supported in Local Testing and skipif is for tests that are blocked on feature or bug fix in development.

File: tests/integ/test_simplifier_suite.py
Comment (sfc-gh-stan) [inline tests/integ/test_simplifier_suite.py:64]: Can this be done using pytest.parameterize?

File: tests/integ/test_column_names.py
Comment (sfc-gh-stan) [reply tests/integ/test_column_names.py:807]: Let's create a followup ticket for this. cc: @sfc-gh-jrose @sfc-gh-aling

File: tests/integ/test_column_names.py
Comment (sfc-gh-stan) [reply tests/integ/test_column_names.py:849]: This is [coercion](https://docs.snowflake.com/en/sql-reference/data-type-conversion).
How did the test work before then?

Comment (sfc-gh-yzou) [reply tests/integ/test_column_names.py:849]: The test doesn't check any result before, I have updated the whole test suite to check the result type and result in this change, and start seeing  mismatches here. It seems that the local testing is probably lacking some test coverage for special values.

File: CHANGELOG.md
Comment (sfc-gh-jdu) [reply CHANGELOG.md]: It seems the first part is not addressed

File: tests/integ/test_column_names.py
Comment (sfc-gh-stan) [inline tests/integ/test_column_names.py:735]: Same issue as below

File: src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py:240]:
```suggestion
    if not isinstance(datatype, _NumericType):
```

File: src/snowflake/snowpark/session.py
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/session.py:653]: You can only send telemetry when it is false for now

File: tests/integ/test_column_names.py
Comment (sfc-gh-yzou) [reply tests/integ/test_column_names.py:807]: created a ticket here https://snowflakecomputing.atlassian.net/browse/SNOW-1524633.  @sfc-gh-aling i assigned it to your for now, feel free to triage it.

File: tests/integ/test_simplifier_suite.py
Comment (sfc-gh-jdu) [inline tests/integ/test_simplifier_suite.py:1243]: you can use https://docs.python.org/3/library/stdtypes.html#str.format instead of replacing strings

File: tests/integ/test_column_names.py
Comment (sfc-gh-stan) [reply tests/integ/test_column_names.py:849]: I created [SNOW-1524637](https://snowflakecomputing.atlassian.net/browse/SNOW-1524637), please put this ticket number wherever you see this issue happen.

**Review by sfc-gh-jdu — APPROVED** [review]
except changelog and local testing change (I didn't review), other parts LGTM

File: tests/integ/test_column_names.py
Comment (sfc-gh-yzou) [reply tests/integ/test_column_names.py:849]: added comment with ticket number

File: tests/integ/test_column_names.py
Comment (sfc-gh-yzou) [reply tests/integ/test_column_names.py:807]: also added the ticket number in comment

File: CHANGELOG.md
Comment (sfc-gh-yzou) [reply CHANGELOG.md]: sorry, updated to refect the user facing expressions

File: tests/integ/test_session.py
Comment (sfc-gh-yzou) [reply tests/integ/test_session.py:687]: updated to xfail, i think I was copying this from other similar tests.

File: tests/integ/test_simplifier_suite.py
Comment (sfc-gh-yzou) [reply tests/integ/test_simplifier_suite.py:64]: yes, but i will need to add a finally to reset the parameter back at the end of each tests, this setup function captures all steps, easier for reuse the code, would prefer to use the fixture here since more than one test point is involved

File: tests/integ/test_column_names.py
Comment (sfc-gh-yzou) [reply tests/integ/test_column_names.py:735]: added comment with ticket number

File: src/snowflake/snowpark/session.py
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/session.py:653]: i actually do want to send telemetry for both values whenever user configures it, the telemetry method I added for this parameter is able to deal with both True and False.

File: tests/integ/test_simplifier_suite.py
Comment (sfc-gh-yzou) [reply tests/integ/test_simplifier_suite.py:1243]: updated

**Review by sfc-gh-stan — APPROVED** [review]
Local Testing test changes LGTM

**Review by sfc-gh-aling — APPROVED** [review]

**Pull request merged**
